Collection of tweaks to Xen versioning information

* Version reduced to $X.$Y.$Z for RPM packaging
* Banner stripped back to Xen-$X.$Y.$Z while keeping -xs$BUILDNUM in the full version
* scmversion updated to split underlying changeset from patchqueue changeset

diff --git a/xen/build.mk b/xen/build.mk
index 758590c68eab..b3886d189b89 100644
--- a/xen/build.mk
+++ b/xen/build.mk
@@ -8,6 +8,7 @@ define cmd_banner
     mv -f $@.tmp $@
 endef
 
+.banner: XEN_FULLVERSION := $$(echo $(XEN_FULLVERSION) | sed "s/-.*//")
 .banner: tools/xen.flf FORCE
 	$(call if_changed,banner)
 
diff --git a/xen/tools/scmversion b/xen/tools/scmversion
index b6ce0498d5a7..54563856bcd4 100755
--- a/xen/tools/scmversion
+++ b/xen/tools/scmversion
@@ -39,6 +39,22 @@ scm_version()
 
 	# Check for git and a git repo.
 	if test -d .git && head=`git rev-parse --verify --short HEAD 2>/dev/null`; then
+
+		if test -d .git/patches/.git; then
+			# Guilt patch queue - Grab the revisions separately
+			for BRANCH in patches qparent patchqueue; do
+				head=`git rev-parse --verify --quiet --short=12 $BRANCH` && break || :
+			done
+
+			if test -n $head; then
+
+				pqhead=`git --git-dir=.git/patches/.git rev-parse --verify --short=12 HEAD`
+
+				printf '%s, pq %s' "$head" "$pqhead"
+				return
+			fi
+		fi
+
 		date=`git show -s --pretty="%ad" HEAD`
 
 		printf '%s %s%s' "$date" git: $head
